### DISCUSSION TITLE
Plan: New Hermes/RNV8 debugger support

### DISCUSSION THREAD

**Discussion by jhen0409** [jhen0409/react-native-debugger#774] [Announcements]
Related: https://github.com/jhen0409/react-native-debugger/issues/573

I'm started experimenting with support for Hermes debugger (it also works on [RNV8](https://github.com/Kudo/react-native-v8)) on RNDebugger. The main purpose is we can keep the same debug experience with this tool.

Until recently I started using JSI and planning to support new architecture in my projects. Flipper is awesome tool to help me debug more deeper issues, but it have some performance issue in my use cases, I wish we had something more lightweight & out-of-the-box tool like RNDebugger.

I'll start the related work at August, these are some key support items:

- DevTools
  - Use devtools-frontend on Electron webview (You can see [the code](https://github.com/facebook/react-native/blob/main/packages/dev-middleware/src/utils/launchChromeDevTools.js) of how React Native launch the chrome devtools)

- Enable open in editor in console ([docs/enable-open-in-editor-in-console.md](https://github.com/jhen0409/react-native-debugger/tree/master/docs/enable-open-in-editor-in-console.md))
  - react-native-v8 devtools support the file link so we can still use this feature, but Hermes seems not

- Console REPL
  - Hermes behaves differently between iOS and Android when using the  console REPL, I can use it on Android but have some weird problems. It is works well with react-native-v8.

- Network tab
  - Expo has great built-in network inspect support on the devtools, we can see if it is can support pure RN projects

- React / Redux (remotedev) / Apollo Client DevTools support
  - Due it not working on WebWorker, we can't do anything script injection, so it will require additional package setup.

- Storage
  - Redux DevTools has a common interface so we can support any libs not limited to state management. I might implement an one for react-native-mmkv by [listeners](https://github.com/mrousavy/react-native-mmkv/blob/master/docs/LISTENERS.md).

The initial version of new debugger implementation will be included in v0.15. Before this, I'll upgrade dependencies & fix some small issues for "old" remote debugger in v0.14.

I'm currently only spending 10~20% of my workday doing it, so any help or advice would be appreciated.

**Comment by AJTJ** [comment]
Thanks for working on this.

This apparently works with a bare React Native app: https://github.com/gusgard/react-native-devsettings but I have not been able to get it working with Expo in managed workflow. Which is a pretty large portion of people building with react-native.

**Reply by jhen0409** [reply]
In my test of Expo 49, you can still use the DevSettings by set `"jsEngine": "jsc"` in `app.json`, but I saw some other problems, maybe it could be about Chromium version or React DevTools. It nothing happened on hermes, I would suggest you to report the issue with Expo if you really want it to support trigger remote debugging for hermes again.

Also, I hope we should focus on the new debugger support, the new architecture will also usher in stability, it is time to get rid of old remote debugging.

**Reply by simonmaquilon** [reply]
It's not neccessary to change the jsEngine, instead enabling Debug mode in the Emulator/Simulator/Device/Expo makes that RND works fine... until your next release.

**Reply by jduhking** [reply]
I’m sorry  if this seems like a dumb question but where is Emulator/Simulator/Device/Expo?

**Comment by ice-cap0** [comment]
any progress?

**Reply by zsomborgellerfi** [reply]
Bump🙌

**Reply by jhen0409** [reply]
There isn't a lot of time available at the moment, and I'm currently focusing less on using the debugger at work (mostly developing for native modules). I hope I'll be motivated to set aside some extra time for that when I switch back to some UI works.

Currently I only do some basic setup on it, you can take a look in the [`new-debugger`](https://github.com/jhen0409/react-native-debugger/tree/new-debugger) branch.

**Reply by RobertoJoseph** [reply]
I have a question and I don't know If it is related to that or not but after I configured my project with eas build and eas update. the option for Remote Js debugger disappeared and only the Open Debugger option is in the developer menu. Even after clicking on Open debugger, It is not open React native debugger but It opens chrome dev tools one, and I could not debug the redux and all the other stuff. @jhen0409

**Reply by eithe** [reply]
@RobertoJoseph Probably not the best place for support questions this thread.

**Comment by RobertoJoseph** [comment]
Any updates?

**Comment by john-mokaied** [comment]


**Reply by dennzimm** [reply]
Just read a little further up the page ...
https://github.com/jhen0409/react-native-debugger/discussions/774#discussioncomment-7007740

**Comment by MarkPan726** [comment]
Maybe this is useful:
import { NativeModules } from 'react-native'

if (__DEV__) {
  NativeModules.DevSettings.setIsDebuggingRemotely(true)
}

**Reply by huynhquangkieu** [reply]
@MarkPan726  Thank for sharing. It work !!!

**Reply by joabgonzalez** [reply]
Thanks! It works for me

**Reply by poinch** [reply]
Where should i place this snippet?
i tried in my App.tsx but when i run my app doesn't load and seems that the debugger is in a loop

**Reply by ruibinch** [reply]
@gusgard has a fantastic package that solves this issue: https://github.com/gusgard/react-native-devsettings

**Reply by poinch** [reply]
i've read that the package does not work in a expo project though

**Reply by MarkPan726** [reply]
The above snippet currently only tests Hermes, the debugger tools support expo by default.

**Reply by kieuhuynhdev** [reply]
I have same issue, react native debugger loop on Android

**Reply by JimTeva** [reply]
thanks @ruibinch for mentioning this and @gusgard for doing the package!!

**Reply by bonnmh** [reply]
it's not working for me :(

**Reply by VisheshG8** [reply]
Thank you so much it helps alot , I spend 1 whole day to find the solution

**Reply by DenisDov** [reply]
This line is the reason for the crash that occurs immediately after the release app opens. In debug mode it's fine.

**Comment by fedorovichandrey** [comment]
Hi [jhen0409](https://github.com/jhen0409), I hope you're doing well! How is it going?

**Reply by dzengi** [reply]
@jhen0409  Hi. Maybe there is some news? Thank you

**Comment by istvan-panczel** [comment]
Ahoy,

For anyone who got frustrated with Hermes and the JS debugger in a React-Native app (I use Expo), here is a quick alternative solution for logging out Redux States (I use Redux Toolkit) while we are waiting for actual dev-friendly tools, I use lodash in my project:

```typescript
const additionalMiddlewaresInDevMode = __DEV__ ? [stateChangeLogger] : [];

const store = configureStore({
  reducer: persistedReducer,
  devTools: false,
  middleware: [saga, ...additionalMiddlewaresInDevMode],
});
```

```typescript
import { isEqual } from 'lodash';
import { AnyAction, Dispatch, Middleware, MiddlewareAPI } from 'redux';

/**
 * Logs the differences between two states
 * This function recursively compares two objects and logs the differences between them.
 *
 * @param prevState - The previous state
 * @param nextState - The next state
 */
function logStateDifferences(prevState: any, nextState: any) {
  const differences: { [key: string]: any } = {};

  // Recursively compare objects for differences
  const compareObjects = (prev: any, next: any, path: string) => {
    Object.keys(next).forEach((key) => {
      const currentPath = path ? `${path}.${key}` : key;

      // Check if the key exists in the previous state
      if (!(key in prev)) {
        differences[currentPath] = {
          a_prev: undefined,
          b_next: next[key],
        };
      }
      // Check if the values are different
      else if (!isEqual(prev[key], next[key])) {
        // If the values are objects, recursively compare them
        if (typeof prev[key] === 'object' && typeof next[key] === 'object') {
          compareObjects(prev[key], next[key], currentPath);
        } else {
          differences[currentPath] = {
            a_prev: prev[key],
            b_next: next[key],
          };
        }
      }
    });
  };

  compareObjects(prevState, nextState, '');

  if (Object.keys(differences).length > 0) {
    console.table(differences);
  } else {
    console.log('%cNo State Differences', 'color: cyan');
  }
}

/**
 * Logs the state before and after the dispatch
 * Use this middleware to log the state before and after the dispatch until we get an actually
 * working Redux Devtools for React-native with Hermes as the JS engine.
 *
 * It groups the logs by the action type and logs the state before and after the dispatch.
 * If the states are identical, it logs that they are identical.
 *
 * @param api - The redux middleware API
 */
const stateChangeLogger: Middleware = (api: MiddlewareAPI) => (next: Dispatch<AnyAction>) => (action: AnyAction) => {
  const previousState = api.getState();
  const result = next(action);
  const afterState = api.getState();

  console.groupCollapsed(`%cAction Dispatched: ${action.type}`, 'color: yellow');
  console.log('%cState Before Dispatch:', 'color: orange', previousState);
  console.log('%cAction:', 'color: yellow', action);
  console.log('%cState After Dispatch:', 'color: orange', afterState);

  if (isEqual(previousState, afterState)) {
    console.log('%cStates are identical', 'color: cyan');
  } else {
    console.groupCollapsed('%cState Differences:', 'color: cyan');
    logStateDifferences(previousState, afterState);
    console.groupEnd();
  }

  console.groupEnd();
  return result; // Return the result of the action dispatch
};

export default stateChangeLogger;
```

An xample of how does it look like on console:
<img width="736" alt="image" src="https://github.com/jhen0409/react-native-debugger/assets/56842295/3fec8c0d-1790-4427-a32d-008531d0f181">

**Reply by andrewgraemebrooks** [reply]
That you so much for this, this is a great solution to use while waiting for some dev-tools. For anyone else who isn't using TypeScript in their project, here is a Javascript version without the Typescript type annotations:
```JS
import { isEqual } from 'lodash';

/**
 * Logs the differences between two states
 * This function recursively compares two objects and logs the differences between them.
 *
 * @param {any} prevState - The previous state
 * @param {any} nextState - The next state
 */
function logStateDifferences(prevState, nextState) {
    const differences = {};

    // Recursively compare objects for differences
    const compareObjects = (prev, next, path) => {
        Object.keys(next).forEach((key) => {
            const currentPath = path ? `${path}.${key}` : key;

            // Check if the key exists in the previous state
            if (typeof prev === 'object' && prev !== null && !(key in prev)) {
                differences[currentPath] = {
                    a_prev: undefined,
                    b_next: next[key],
                };
            }
            // Check if the values are different
            else if (!isEqual(prev[key], next[key])) {
                // If the values are objects, recursively compare them
                if (
                    prev[key] !== null &&
                    typeof prev[key] === 'object' &&
                    next[key] !== null &&
                    typeof next[key] === 'object'
                ) {
                    compareObjects(prev[key], next[key], currentPath);
                } else {
                    differences[currentPath] = {
                        a_prev: prev[key],
                        b_next: next[key],
                    };
                }
            }
        });
    };

    compareObjects(prevState, nextState, '');

    if (Object.keys(differences).length > 0) {
        console.table(differences);
    } else {
        console.log('%cNo State Differences', 'color: cyan');
    }
}

/**
 * Logs the state before and after the dispatch
 * Use this middleware to log the state before and after the dispatch until we get an actually
 * working Redux Devtools for React-native with Hermes as the JS engine.
 *
 * It groups the logs by the action type and logs the state before and after the dispatch.
 * If the states are identical, it logs that they are identical.
 *
 * @param api - The redux middleware API
 */
const stateChangeLogger = (api) => (next) => (action) => {
    const previousState = api.getState();
    const result = next(action);
    const afterState = api.getState();

    console.groupCollapsed(`%cAction Dispatched: ${action.type}`, 'color: yellow');
    console.log('%cState Before Dispatch:', 'color: orange', previousState);
    console.log('%cAction:', 'color: yellow', action);
    console.log('%cState After Dispatch:', 'color: orange', afterState);

    if (isEqual(previousState, afterState)) {
        console.log('%cStates are identical', 'color: cyan');
    } else {
        console.groupCollapsed('%cState Differences:', 'color: cyan');
        logStateDifferences(previousState, afterState);
        console.groupEnd();
    }

    console.groupEnd();
    return result; // Return the result of the action dispatch
};

export default stateChangeLogger;
```

**Comment by saxenanickk** [comment]
@jhen0409 Any progress on this?

**Comment by liquidvisual** [comment]
For anyone on Expo trying to get Redux DevTools to work, there's a temporary solution using a plugin. [See here](https://github.com/reduxjs/redux-devtools/issues/1533#issuecomment-2146370963)

**Comment by losh11** [comment]
Is anyone interested in making a front end for @istvan-panczel's hook? Instead of console.logging the diff-ed redux state, it's sent to an observing front end app that makes it look all pretty like react-native-debugger? 
TBH the loss of being able to see the current redux state + state changes has been a massive pain when debugging.

**Reply by istvan-panczel** [reply]
I recently had a similar idea!

In addition to using the state logger, I implemented HTTP interceptors to log every request and response, including headers, data, and timestamps. I also added loggers at the DB layer and subscribed to screen changes so that each time a screen is loaded, its name and route parameters are logged.

With these enhancements, I feel confident in my ability to debug any issue in my current contract project. By navigating the app and interacting with it, I can easily see detailed logs. Each type of log has distinct colors and formatting, is expandable, and can be inspected thoroughly. These logging points also facilitated the integration of Sentry into the project.

I'm considering developing a separate application to run locally, which would replace console logging with simple HTTP requests to this app. The idea is to have a small database to store all logs and a user-friendly UI to visualize the state, changes, and activities at the DB and network layers. Although I haven't extensively used the "time-travel" features of the Redux debugger, implementing a feature to apply a specific state to the store of your app at any action point (before or after) seems quite feasible.

**Reply by aliffazfar** [reply]
Can checkout this [774#discussioncomment-10095554](https://github.com/jhen0409/react-native-debugger/discussions/774#discussioncomment-10095554) if you miss the redux time-travel debugging on Hermes

**Comment by sigaits** [comment]
Since I updated to Expo 50, I have had this problem: if I reload the app while the dev-tools is running, the app crashes on the simulator (as in quits and I have to relaunch it). It only started happening since Expo 50 but it's been a while since I updated and still having the same problem. Is anyone else having this issue?

**Reply by DePavlenko** [reply]
it might be connected to react-native-reanimated v3.0 and higher
https://github.com/software-mansion/react-native-reanimated/issues/4783

**Comment by misbell** [comment]
I guess this project is dead in the water, if it can't support Hermes, and there are not updates about it. Too bad, I used it all the time.

**Comment by Ahmed-M-Osman1** [comment]
Any update guys?

**Comment by aliffazfar** [comment]
For those who miss Redux DevTools for time-travel debugging and live editing, I’ve managed to make it work for React Native with the Hermes engine. 

Checkout my repo: https://github.com/aliffazfar/redux-devtools

![Screenshot 2024-07-19 at 5 33 38 PM](https://github.com/user-attachments/assets/39c9ace8-473f-4319-a0e2-7db9cde896e8)

The only thing we miss is the network inspector in Chrome, like in the old days. However, we can move on with [Reactotron](https://docs.infinite.red/reactotron/quick-start/react-native), which works really great for debugging the network, AsyncStorage, and logs.

I have tested and can confirm that [RocketSim](https://www.rocketsim.app/) also work great with React Native to inspect the network. Check this out: https://x.com/AliffAzfar/status/1814121783303569838

**Reply by dariusz-rosiek-branchspace** [reply]
can someone confirm that this work correctly ?

**Reply by rzfzr** [reply]
Does it work with expo?

**Reply by losh11** [reply]
hey @aliffazfar I'd like to try your fork of redux-devtools. Are you planning to maintain it? If so, can you enable github issues on the repo? Also do you have documentation for RTK (redux toolkit) as the redux documentation strongly reccomends using RTK over the older way.

**Reply by eithe** [reply]
If you guys haven't seen it you should take a look at https://www.youtube.com/watch?v=b48Lax2-jOQ&t=4350s which was presented a few weeks ago. They are hoping to add redux support for it later, but it will be available without it from 0.76.

**Reply by aliffazfar** [reply]
> hey @aliffazfar I'd like to try your fork of redux-devtools. Are you planning to maintain it? If so, can you enable github issues on the repo? Also do you have documentation for RTK (redux toolkit) as the redux documentation strongly reccomends using RTK over the older way.

You can use official documentation to debug RTK. My fork was only enabling it for those who wanted to debug with hermes engine.

**Reply by aliffazfar** [reply]
> Does it work with expo?

Expo has much better debugger with hermes engines.

**Comment by c-goettert** [comment]
I just want my redux debug tools back. Bring em baack please 😭

**Comment by Matthew-Jimenez** [comment]
We should really add https://github.com/gusgard/react-native-devsettings to the readme.

**Comment by herujest** [comment]
it's 2025 any update?

**Comment by TonyFTannous-Byd** [comment]
did not work with the new architecture (react native 0.76.6), react native debugger app not working 
any news!!!

**Reply by gusgard** [reply]
You should use this now 

https://reactnative.dev/docs/react-native-devtools

**Reply by bfd-paul-vicknair** [reply]
As far as I can tell, that doesn't handle the redux or network pieces.  Otherwise it appears to be pretty great.

**Comment by TonyFTannous-Byd** [comment]
@gusgard @bfd-paul-vicknair 
Hi
i am using it, but i need react native debugger app to handle redux or network. but it ok i replaced it by reactotron

**Reply by kaanmertkoc** [reply]
reactotron is so worse. UI/UX is terrible.

**Reply by sLurPPPeRsTaR** [reply]
> reactotron is so worse. UI/UX is terrible.

the only thing you can rely on that debugger is storybook the rest is bad experience

**Comment by sLurPPPeRsTaR** [comment]
react native debugger unreplaceable

**Comment by nikilarigela** [comment]
Hey @jhen0409 any update on new dubgger?

**Comment by ferchomuri** [comment]
Can I help u? @jhen0409

**Comment by manish-mahto-982** [comment]
why network section is not showing in react-native devtools
<img width="781" height="32" alt="Screenshot 2025-10-09 at 10 49 12 AM" src="https://github.com/user-attachments/assets/286fce3f-a71c-4b58-886b-136276bc9c5a" />

**Reply by liamjones** [reply]
It's not implemented yet: https://github.com/react-native-community/discussions-and-proposals/discussions/819

<img width="1025" height="360" alt="image" src="https://github.com/user-attachments/assets/918a4fea-7293-4180-98c1-189b58647db2" />

**Comment by CodeByRahulSaini** [comment]
Hello all,
Looks like the CallStack team solved it for all of us and built Rozenite which works with existing React Native Devtools — and a plugin framework that brings a clean, extensible plugin system in React Native DevTools.

- Official plugins already available: Redux DevTools, TanStack Query, network activity monitoring, performance marks & more.
- Works with Expo + bare RN, and lets you build custom plugins for your own use cases.
	
Surprisingly, most people still don’t know this exists — worth checking out!

🔗 https://www.rozenite.dev/
🎥 https://www.youtube.com/watch?v=J0rXmTIGGRk

Would love to know what the community thinks about this.

**Reply by safin-intangles** [reply]
Yes but there are still issues. I recently set it up and ended up wasting time as it wasnt working with my current workflow using saga, dynamic-redux-modules.

**Reply by CodeByRahulSaini** [reply]
That looks like a special use case. Feel free to add an issue here -> https://github.com/callstackincubator/rozenite

**Reply by safin-intangles** [reply]
Have already added long back, no response yet. https://github.com/callstackincubator/rozenite/issues/160

**Reply by Zergus-s** [reply]
God bless you for this info! Suffered with Tanstack Query and Axios debugging on Expo, now with Rozenite it works perfectly both network and cache!
Much better than Reactotron

**Comment by sharanagouda** [comment]
temporary work around

<img width="1296" height="898" alt="Screenshot 2026-06-03 at 10 00 40 AM" src="https://github.com/user-attachments/assets/471bad9b-6fae-4298-b2e1-5ad1c65c788f" />
<img width="1301" height="903" alt="Screenshot 2026-06-03 at 10 01 18 AM" src="https://github.com/user-attachments/assets/a270536c-af6f-4870-ab4e-89d51977c5e2" />
<img width="1303" height="901" alt="Screenshot 2026-06-03 at 10 01 35 AM" src="https://github.com/user-attachments/assets/e9f872d4-3a9b-4626-9f2d-7fdb2f907716" />

🚀 Built ReactoRadar — an open-source React Native Debugger for RN 0.74+ / Hermes / New Architecture

The old React Native Debugger doesn't support Hermes. Flipper is deprecated. So I built a modern replacement from scratch.

What it does:

• Console — collapsible object trees, multi-select level filters, Cmd+F find in logs, Redux actions inline
• Network — Chrome DevTools-style with resizable/sortable columns, Copy as cURL, throttling (3G/Offline)
• Redux — time travel with store diff showing exactly which keys changed (old → new values)
• GA4 Events — intercepts ALL Firebase Analytics methods, event list with parameters, summary chips
• AsyncStorage — live key/value browser with search
• 9 Themes — Dark, Light, Monokai, Dracula, Solarized, Nord, GitHub Dark, One Dark

2 commands to get started:

npx reactoradar setup    ← from your RN project
npx reactoradar           ← launch the debugger

That's it. No config, no native modules. Auto-detects your platform, patches index.js, wires Redux, intercepts Firebase Analytics.

Also available as a .dmg download for macOS.

📦 npm: https://www.npmjs.com/package/reactoradar
💻 GitHub: [GitHub - sharanagouda/react-native-debugger: Standalone macOS debugger for React Native 0.74+ with H…](https://github.com/sharanagouda/react-native-debugger)
📥 Download: [Releases · sharanagouda/react-native-debugger](https://github.com/sharanagouda/react-native-debugger/releases)

Contributions welcome — looking for help with Windows/Linux support and profiling tools.

Give it a try and share your feedback! 🙌

#ReactNative #JavaScript #MobileDevelopment #OpenSource #DevTools #Debugging #Hermes #iOS #Android #Firebase #GA4 #Redux

**Comment by istvan-panczel** [comment]
At this point the whole thing is solved now.

I upgraded the app to expo 56 with all its dependencies and the official plugin works. With latest RN, Hermes, etc.

https://docs.expo.dev/debugging/devtools-plugins/

https://github.com/reduxjs/redux-devtools

**Comment by maahibhama** [comment]
I’ve been building **PulseRN**, an open-source, local-first desktop debugger for React Native.

It brings console logs, network requests, Redux actions, navigation events, performance signals, storage, and errors into one chronological timeline. The idea is to help developers understand the complete sequence behind a bug instead of switching between several disconnected tools.

It also includes:

- Hermes debugging with original TypeScript sources
- Breakpoints, scopes, watches, stepping, and evaluation
- Expo development-build and bare React Native support
- Configurable redaction and bounded local storage
- MCP integration for evidence-backed debugging with Codex, Claude, Cursor, and other compatible clients
- Desktop builds for macOS, Windows, and Linux

GitHub: https://github.com/maahibhama/PulseRN  
Documentation: https://maahibhama.github.io/PulseRN-Site/

I’d really appreciate feedback from React Native developers:

- Does the unified timeline solve a problem you experience?
- Which debugging integration would be most useful to you?
- Is the installation and SDK setup clear?
- What would prevent you from trying this on a real project?

PulseRN is still actively evolving, so honest feedback, bug reports, and contributions are very welcome.
